### PR TITLE
MAID-2856 - fix garbage collected callback delegates in C#

### PR DESCRIPTION
**Pull request by nbaksalyar**
Fixes the issue in the manual part of C# bindings for SAFE Client Libs.
Also see: https://github.com/maidsafe/safe_bindgen/pull/13

### CODE REVIEW

File: safe_app/resources/AppBindings.Manual.cs [hunk 1]
```diff
@@ -67,12 +71,12 @@ public Task<App> AppUnregistered(byte[] bootstrapConfig, Action oDisconnectNotif
...
-                               OnDecodeIpcMsgAuthCb,
-                               OnDecodeIpcMsgUnregisteredCb,
-                               OnDecodeIpcMsgContainersCb,
-                               OnDecodeIpcMsgShareMdataCb,
-                               OnDecodeIpcMsgRevokedCb,
-                               OnDecodeIpcMsgErrCb);
+                               DelegateOnDecodeIpcMsgAuthCb,
+                               DelegateOnDecodeIpcMsgUnregisteredCb,
+                               DelegateOnDecodeIpcMsgContainersCb,
+                               DelegateOnDecodeIpcMsgShareMdataCb,
```
Comment (ravinderjangra) [inline safe_app/resources/AppBindings.Manual.cs]: Different Delegate name(DelegateOnDecodeIpcMsgShareMDataCb) in line 126.

**Comment by nbaksalyar** [comment]
@oetyng,
>Is this use of a static reference thread safe?
>(Does it need to be?)

As per [MSDN](https://msdn.microsoft.com/en-us/library/system.delegate(v=vs.110).aspx#Anchor_10), it is thread safe. Although it doesn't say anything specific about `private static` fields, I don't think it can make the delegates non-TS.

File: safe_authenticator/resources/SafeAuth.AuthBindings/AuthBindings.Manual.cs [hunk 2]
```diff
@@ -54,6 +54,8 @@ internal partial class AuthBindings {
 
       action(Marshal.PtrToStructure<FfiResult>(result), app, GCHandle.FromIntPtr(userData));
     }
+    private static FfiOnAuthenticatorCreateCb DelegateOnAuthenticatorCreateCb = OnAuthenticatorCreateCb;
```
Comment (ravinderjangra) [inline safe_authenticator/resources/SafeAuth.AuthBindings/AuthBindings.Manual.cs]: Delete FfiOnAuthenticatorCreateCb not declared. same issue as SafeApp.Manual.cs

File: safe_authenticator/resources/SafeAuth.AuthBindings/AuthBindings.Manual.cs [hunk 3]
```diff
@@ -62,6 +64,7 @@ internal partial class AuthBindings {
 
       action();
     }
+    private static FfiOnAuthenticatorDisconnectCb DelegateOnAuthenticatorDisconnectCb = OnAuthenticatorDisconnectCb;
```
Comment (ravinderjangra) [inline safe_authenticator/resources/SafeAuth.AuthBindings/AuthBindings.Manual.cs]: Delegate not Decleared

File: safe_authenticator/resources/SafeAuth.AuthBindings/AuthBindings.Manual.cs [hunk 4]
```diff
@@ -70,6 +73,7 @@ internal partial class AuthBindings {
       var tcs = BindingUtils.FromHandlePtr<TaskCompletionSource<IpcReq>>(userData);
       tcs.SetResult(new AuthIpcReq(reqId, new AuthReq(Marshal.PtrToStructure<AuthReqNative>(authReq))));
     }
+    private static FfiOnDecodeIpcReqAuthCb DelegateOnDecodeIpcReqAuthCb = OnDecodeIpcReqAuthCb;
```
Comment (ravinderjangra) [inline safe_authenticator/resources/SafeAuth.AuthBindings/AuthBindings.Manual.cs]: Delegate not Decleared

**Pull request merged**